### ISSUE TITLE
pipelined requests bug

### ISSUE THREAD

**Issue by rickynelson1939** [warmcat/libwebsockets#946]
if pipelined requests are sent in glued chunk larger than 2048 bytes then connection is stopped processing and warning "Duplicated method" is given

**Comment by rickynelson1939** [comment]
damn, it's hard to reproduce, here's binary data that works for me
[dump.zip](https://github.com/warmcat/libwebsockets/files/1100177/dump.zip)

**Comment by rickynelson1939** [comment]
damn it, even log level change cause this bug to change its behaviour...

**Comment by rickynelson1939** [comment]
[log.txt](https://github.com/warmcat/libwebsockets/files/1100533/log.txt)

**Comment by rickynelson1939** [comment]
well, judging by the log, next pipelined request came before file, issued by previous request, was completely send -> "lws_http_transaction_completed" wasn't called, fragments wasn't reset -> next pipelined request was read on top of previous request headers

**Comment by lws-team** [comment]
Does the patch above help?

**Comment by rickynelson1939** [comment]
No more "Duplicated method" warning, but it gets strange, when log level is at 7 or less there is "forbidding on uri sanitation" warning given, but when I'm trying to log it at least on debug level (15) then there is no errors and traffic is processing a lot slower, so it's hard to tell what is happening exactly...

**Comment by lws-team** [comment]
This (which goes out as one pipelined request)

```
$ curl http://localhost:7681/favicon.ico  http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico  http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico  http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico  http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico http://localhost:7681/favicon.ico > /dev/null
```

doesn't have any problem here.  It didn't have any problem before the last patch either, FWIW.

100 x pipelined `http://localhost:7681/libwebsockets.org-logo.png` with larger payload is also OK.  Doing 100 x curl favicon.ico from a machine on another continent doesn't show up any problem.

**Comment by rickynelson1939** [comment]
I actually testing this just with a buffer filled with repeated single http request sent via "send" in blocking loop, if that's matter. The fact that this bug is not reproducible even on my machine if log level is high enough should tell about something, I guess it takes some special condition, e.g specific case of fragmentation or something like that.

**Comment by rickynelson1939** [comment]
Maybe if I dump every result of call to a recv and then repeat this exact data and call sequence, but with higher log level it will reveal something. Or maybe you have some better ideas?

**Comment by lws-team** [comment]
I can't reproduce the problem.... it doesn't mean the problem doesn't exist.

Using your test packet doesn't show up any problem (with default log level) here either.

```
$ cat ~/dump.bin | nc localhost 7681
```

That sends the whole test packet the first time, still no problem.

If you can casually reproduce the issue, you will have to stick some debugging around `forbid:` in parsers.c: 953, find out who jumped there and what made him feel that was a good idea.  That logging will happen after the problem comes so it shouldn't affect reproducibility on your end.

**Comment by rickynelson1939** [comment]
Okay, I'll try some of that right now.

**Comment by rickynelson1939** [comment]
it jumps to forbid from parsers.c: 833, and I actually noticed, all frag_index'es are zero

**Comment by lws-team** [comment]
Might be interesting to instrument server.c::1288 lws_handshake_server() who calls lws_parse()... if he errors out, dump a copy of *buf contents from entry and see if what it tries to parse is sensible.

**Comment by rickynelson1939** [comment]
0x642e32 "1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r"...

**Comment by lws-team** [comment]
It's more useful if you dump the *buf as it was **on entry** to lws_handshake_server(), if lws_parse() fails.  Then we can see what it was trying to parse.

**Comment by rickynelson1939** [comment]
so you mean obuf then?
0x642e31 ".1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n"...

**Comment by lws-team** [comment]
So it was given just the one char '.'... we also know if it got to parsers.c: 833 it did not have any method yet... either something reset the ah while it was still in use or we start at an unreasonable place in the rx.

Will the problem keep coming if you add logging to _lws_header_table_reset() parsers.c:64, and some log to say we entered lws_handshake_server(), so we can understand the ordering a bit?

**Comment by rickynelson1939** [comment]
wierd thing, if I used lwsl_warn - there was no error, but after I used fprintf - it did triggered forbid
[log.txt](https://github.com/warmcat/libwebsockets/files/1104278/log.txt)

**Comment by lws-team** [comment]
I tried changing the log emit function to the default stderr (test-server.c 340 -> 	lws_set_log_level(debug_level, NULL);) , the problem didn't start appearing.  Can you see if that lets you dump logs?

Basically the successful ones have this pattern

```
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
```

but the failing one is only

```
lws_handshake_server
```

It should be OK since already one reset before it was called.  But it suggests that's part of a different behaviour that might be the real problem.

**Comment by rickynelson1939** [comment]
Hmm, no, it stop working when I do that. I even tried replace every lwsl_* so it define plain fprintf, but no luck... I think it's time to dump recv

**Comment by rickynelson1939** [comment]
[recv_dump.zip](https://github.com/warmcat/libwebsockets/files/1104357/recv_dump.zip)

**Comment by lws-team** [comment]
What exactly am I looking at?

**Comment by rickynelson1939** [comment]
this is data exactly as it was received from socket when bug was triggered, from start to the very moment of breakpoint, now I just emulate each and every result of a recv call exactly as it was in previous session, but now with log level 65535, and I'll be damned, it worked
[log.txt](https://github.com/warmcat/libwebsockets/files/1104374/log.txt)

**Comment by lws-team** [comment]
Still... it's something to do with something... packet 76 ends with

```
GET /favicon.ico HTTP/1
```

and then naturally packet 77 begins with

```
.1
```

so we learned we blow on fragmentation of the first line... lws has bytewise parsers for everything, it should not matter to it how the stuff is fragmented.

Is it possible the logging sensitivity is to do with the kernel having time to coalesce the two packets?  Can you check that?

**Comment by lws-team** [comment]
Trying to duplicate this standalone, as expected it is immune to fragmentation.  In one terminal

```
 $  nc -l -k 7680 | nc localhost 7681
```

then in another (quickly before it timesout)

```
 $ echo -n "GET /favicon.ico HTTP/1" | nc localhost 7680 ; echo -n -e ".1\xa\xd" | nc localhost 7680
```

gets me

```
[2017/06/27 14:57:17:0485] DEBUG: lws_read: incoming len 23  state 0
[2017/06/27 14:57:17:0485] PARSER: issuing 23 bytes to parser
[2017/06/27 14:57:17:0485] PARSER: WSI_TOKEN_NAME_PART 'G' (mode=0)
[2017/06/27 14:57:17:0485] PARSER: WSI_TOKEN_NAME_PART 'E' (mode=0)
[2017/06/27 14:57:17:0485] PARSER: WSI_TOKEN_NAME_PART 'T' (mode=0)
[2017/06/27 14:57:17:0485] PARSER: WSI_TOKEN_NAME_PART ' ' (mode=0)
[2017/06/27 14:57:17:0485] PARSER: known hdr 0
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) '/'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'f'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'a'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'v'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'i'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'c'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'o'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'n'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) '.'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'i'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'c'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) 'o'
[2017/06/27 14:57:17:0485] PARSER: WSI_TOK_(0) ' '
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) 'H'
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) 'T'
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) 'T'
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) 'P'
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) '/'
[2017/06/27 14:57:17:0486] PARSER: WSI_TOK_(15) '1'
[2017/06/27 14:57:17:0486] DEBUG: lws_read: thinks we have used 0
[2017/06/27 14:57:17:0486] INFO: lws_read: read_ok, used 23
[2017/06/27 14:57:17:0486] DEBUG: lws_server_socket_service: wsi 0x258f070: ah read rxpos 23, rxlen 23
[2017/06/27 14:57:17:0615] DEBUG: lws_read: incoming len 4  state 0
[2017/06/27 14:57:17:0616] PARSER: issuing 4 bytes to parser
[2017/06/27 14:57:17:0616] PARSER: WSI_TOK_(15) '.'
[2017/06/27 14:57:17:0616] PARSER: WSI_TOK_(15) '1'
[2017/06/27 14:57:17:0616] PARSER: WSI_TOK_(15) '
'
'2017/06/27 14:57:17:0616] PARSER: WSI_TOK_(15) '
[2017/06/27 14:57:17:0616] PARSER: *
[2017/06/27 14:57:17:0616] DEBUG: lws_read: thinks we have used 0
[2017/06/27 14:57:17:0616] INFO: lws_read: read_ok, used 4
[2017/06/27 14:57:17:0616] DEBUG: lws_server_socket_service: wsi 0x258f070: ah read rxpos 4, rxlen 4
[2017/06/27 14:57:20:0198] NOTICE: wsi 0x258f070: TIMEDOUT WAITING on 3 (did hdr 0, ah 0x23b1480, wl 0, pfd events 1) 1498546640 vs 1498546639
```

Without the logging he just shows timeout, ie, no failure at parse.

I'll try fake up the same pipelined as well.

**Comment by rickynelson1939** [comment]
Hmm, I'll try, but how am I supposed to do this exactly?
here's a log with a log level 65535 and input from real socket, no forbid was triggered
[log.txt.zip](https://github.com/warmcat/libwebsockets/files/1104397/log.txt.zip)

**Comment by lws-team** [comment]
I just mean do the same recv info dump you did before, but with logging enabled... are the 76th and 77th packet still fragmented in the same place?

**Comment by rickynelson1939** [comment]
no, it seems at some point something went different
[recv_dump_no_bug.zip](https://github.com/warmcat/libwebsockets/files/1104417/recv_dump_no_bug.zip)

**Comment by rickynelson1939** [comment]
wait a minute, what a hell, there's a bunch of zeros at the end of first file

**Comment by lws-team** [comment]
FWIW I found I can spam endless amounts of fragmented and unfragmented requests on one connection with no problems.

```
 $ while [ 1 ] ; do echo -n -e "GET /favicon.ico HTTP/1.1\x0d\x0a\x0d\x0aGET /favicon.ico HTTP/1" | nc localhost 7680 ; echo -n -e ".1\x0d\x0a\x0d\x0a" | nc localhost 7680 ; done
```

**Comment by lws-team** [comment]
> wait a minute, what a hell, there's a bunch of zeros at the end of first file

So there is... the file size is 2KB though

**Comment by rickynelson1939** [comment]
damn, I used wrong variable in dump code, I redump both cases right now

**Comment by rickynelson1939** [comment]
[recv_dump_with_bug.zip](https://github.com/warmcat/libwebsockets/files/1104434/recv_dump_with_bug.zip)
[recv_dump_no_bug.zip](https://github.com/warmcat/libwebsockets/files/1104435/recv_dump_no_bug.zip)

**Comment by lws-team** [comment]
Hmmm in the bugging dump the earlier pkt #82 also begins with ".1" but that has no problem.  Presumably it blows on pkt 157, the last one, which begins again with ".1"

pkts 82 and 157 are the same in both dumps.  So it's not to do with timing or size of the rx.

**Comment by rickynelson1939** [comment]
and the horrible truth is, they are absolute same on every file, but one of them did trigger bug and the other is not...

**Comment by lws-team** [comment]
If it's not timing it's usually stack.

**Comment by rickynelson1939** [comment]
And yet it has pattern, obuf started with '.' every time. So it's not random, I suppose. What about stack? It gets corrupted? or somewhere is uninitialized variable that derive its value from dirty stack? Without log this error is incredibly stable, each time.

**Comment by lws-team** [comment]
The stack contents are not actually random either... it's not just inherited data but also pointer writes behind stack buffers writing on caller's data for example, which is only noticed after it pops back up to the parent and it uses it. Other than this suspicious .1 business it acts very nondeterministic.... if there's only one thread and it's not network timing then that's actually quite hard to achieve.

There's also the fact I have not been able to trigger it at all.... it could be specific to platform or toolchain (I am on x86_64 Fedora 26 / gcc 7.1.1).

**Comment by rickynelson1939** [comment]
Ubuntu 16.04.2 LTS
gcc version 5.4.0 20160609 (Ubuntu 5.4.0-6ubuntu1~16.04.4)

**Comment by rickynelson1939** [comment]
usleep(20 * 1000); before recv cures this as well

**Comment by rickynelson1939** [comment]
maybe it got something to do with the way my lws is configured, I leave it here, just in case:
```
-DCMAKE_BUILD_TYPE=DEBUG -DLIB_SUFFIX=64 -DCMAKE_CXX_FLAGS=-m64 -DCMAKE_C_FLAGS=-m64 \
-DLWS_WITH_SHARED=OFF \
-DLWS_WITH_SSL=OFF \
-DLWS_USE_BUNDLED_ZLIB_DEFAULT=OFF \
-DLWS_WITH_ZLIB=OFF \
-DLWS_WITHOUT_CLIENT=ON \
-DLWS_WITHOUT_TEST_SERVER_EXTPOLL=ON \
-DLWS_WITHOUT_TEST_PING=ON \
-DLWS_WITHOUT_TEST_ECHO=ON \
-DLWS_WITHOUT_TEST_CLIENT=ON \
-DLWS_WITHOUT_TEST_FRAGGLE=ON \
-DLWS_WITHOUT_EXTENSIONS=ON \
-DLWS_WITH_RANGES=OFF \
-DLWS_WITH_ZIP_FOPS=0
```

**Comment by lws-team** [comment]
Those options (including the m64) don't change anything here.  Is your platform x86_64 or 32-bit though?

Did you remove this abortive patch from Sunday about early ah detach?  It's removed in master.

**Comment by rickynelson1939** [comment]
I just cloned master branch when we started. Ubuntu is 64 bit

**Comment by lws-team** [comment]
Reviewing what we found so far, the evidence of the problem is there for an extra call to lws_header_table_reset().

When we get the headers split between two packets it's supposed to do this.... this is taken from your log line 6148

```
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
_lws_header_table_reset
lws_handshake_server
lws_handshake_server
```

The first three lines do the whole headers in one, but then on the last 3, at first it clears down the ah and starts to parse the last part of the current packet.  Then it runs out, and the last line is it resuming the parse from the next packet.  There is no problem.

But just before it blows the pattern is (line 6226 in your log)

```
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
lws_handshake_server
```

There is an extra ah reset inbetween, which kills the parser state.  When we resume on the next packet, the parser as gone back to zero and is looking for the method, and gets "." which fails out since it doesn't match any method.

The only non-client guys who reset the ah are

 - `lws_header_table_attach()`, if you call it when you already have an ah, or you get it attached for the first time
 - `lws_header_table_detach()` if he decides to give it to someone on the waiting list
 - `lws_http_transaction_completed()` 

Can you add logs to those three and try the test logging lws_handshake_server and _lws_header_table_reset again?  The first two shouldn't get called past the start and end, but `ws_http_transaction_completed()` might be too chatty for the bug.  If so you can try swallowing the first n of them until we just get ones near the bug.

**Comment by rickynelson1939** [comment]
[log.txt](https://github.com/warmcat/libwebsockets/files/1104728/log.txt)

**Comment by lws-team** [comment]
Thanks... that gave me an idea of the root cause, but I have to do other stuff now.  I'll study it later or tomorrow and reply then.

**Comment by rickynelson1939** [comment]
Okay. And thank you too, this was interesting to investigate :)
I need to go to sleep soon anyway.

**Comment by lws-team** [comment]
Please update against master and see if any change.

**Comment by rickynelson1939** [comment]
I cloned master branch again, since I mutilated too much code yesterday, and... "forbidding on uri sanitation" again

**Comment by lws-team** [comment]
It's still coming at the same point, ie, when we get a packet starting ".1"?

I checked I can spam that endlessly still without problems.

If it's still coming on ".1", how about the test from yesterday that ends with

```
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
_lws_header_table_reset
lws_handshake_server
_lws_header_table_reset
lws_handshake_server
```

it's still the same?

**Comment by rickynelson1939** [comment]
Yeap, .1 it is.
[log.txt](https://github.com/warmcat/libwebsockets/files/1107566/log.txt)

**Comment by lws-team** [comment]
Is your libwebsockets-test-server also coming out of the current lws build or you cloned it from before?

**Comment by rickynelson1939** [comment]
```
rm -r -f libwebsockets 
git clone https://github.com/warmcat/libwebsockets.git -b master
```

**Comment by rickynelson1939** [comment]
It seems that I can fprintf some key functions and bug still does appear, so maybe this can be used to debugging advantage

**Comment by lws-team** [comment]
It's good, but I was asking about the test server.  If its sources are the ones built by the current lws, that's good since there are changes in there.  If you copied the previous test server sources out to one side as the start of your own app, that would also need to be updated now.,, that's what I was checking.

```
lws_handshake_server
lwsts[7497]: lws_http_transaction_completed: wsi 0x62beb0
lwsts[7497]: lws_http_transaction_completed: wsi 0x62beb0
_lws_header_table_reset
lws_handshake_server
lwsts[7497]:  forbidding on uri sanitation
```

So the first `lws_handshake_server` should be using up to the end of the penultimate packet.  After he runs dry, he should not have a completed set of headers,  So not only should nothing be under the impression any transaction completed but the `lws_http_transaction_completed()` function should bail on the basis no transaction could be completed if the http header parsing is incomplete.

After that we think we understand, the ah got reset partway through and then the parsing state is lost and we forbid it immediately.

Can you maybe dump *obuf again in lws_handshake_server, or 80 chars or whatever of it or so?

**Comment by lws-team** [comment]
There's also a flag `wsi->hdr_parsing_completed` which should be zeroed when we reset the ah and only set when we finished a set of http headers.  `lws_http_transaction_completed()` just exits cleanly if it gets called when `wsi->hdr_parsing_completed` is 0.  Maybe dumping that too will help towards a clue.

**Comment by rickynelson1939** [comment]
No, I did not change a line after clone.

I periodically check obuf and it's strangely have always the same value.

Name : wsi->hdr_parsing_completed
	Decimal:0

Name : obuf
	Details:0x63ce31 ".1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nHost: 192.168.137.27\r\n"...

**Comment by lws-team** [comment]
Couple of ideas... how about add a fake header line `n: 123` and increment the number each time.  Lws will ignore it for parsing purposes but it might be useful to confirm we are where we think we are in the buffer.

It's a bit painful but we could do with instrument every call to `lws_http_transaction_completed()` with a unique log just before it so we know where those last two are called from.

**Comment by rickynelson1939** [comment]
wow, with n: 123 being added:
Name : obuf
	Details:0x63ce31 "3\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nn: 123\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nn: 123\r\nHost: 192.168.137.27\r\n\r\nGET /fa"...

**Comment by lws-team** [comment]
Ah I meant like n: 1, n: 2, n:3 etc so we can see where we are in the stream

**Comment by lws-team** [comment]
Oh.... I get it..... we die with the last packet on "3" now?

**Comment by rickynelson1939** [comment]
no, I forget about incrementing counter, I'll do it now

**Comment by rickynelson1939** [comment]
but it will take a minute, I'm kinda just filling the buffer at the start, so I have to change a bit of code...

**Comment by lws-team** [comment]
No worries. The annotation of the transaction completed callers is also still interesting.

**Comment by rickynelson1939** [comment]
It started taking more time, like ~4 seconds instead of immediately, maybe because requests start taking time to compile into glued piece every time, but it does bug eventually

Name : obuf
	Details:0x63ce31 "icon.ico HTTP/1.1\r\nConnection: keep-alive\r\nn: 11978\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection: keep-alive\r\nn: 11979\r\nHost: 192.168.137.27\r\n\r\nGET /favicon.ico HTTP/1.1\r\nConnection:"...

[log.txt](https://github.com/warmcat/libwebsockets/files/1107647/log.txt)

**Comment by rickynelson1939** [comment]
seems like the number is different each time, but once I got same exact number twice in a row, but no luck after that

**Comment by lws-team** [comment]
I think this is still looking at a 'secondary effect' of the real problem.

The real problem happens earlier, before the two transaction completed calls.

So it might help to know which calls they are.  There are 11 relevant ones on the lib and one in the test server.

**Comment by rickynelson1939** [comment]
so I need to add log to every one of them, yes?

**Comment by lws-team** [comment]
Yeah or half of them and see if they are the ones, then bother to do the other half.

**Comment by rickynelson1939** [comment]
[log.txt](https://github.com/warmcat/libwebsockets/files/1107715/log.txt)

**Comment by lws-team** [comment]
I just pushed an additional patch https://github.com/warmcat/libwebsockets/commit/a637d8f41f520d33356d16b9946d58b0073fdc12  does it change anything?

**Comment by rickynelson1939** [comment]
hmmm, it's kinda shameful to ask, but I'm not using git so much, I can do pull but it deletes log I added, is that okay?

**Comment by lws-team** [comment]
To follow master the quickest way is like this

1) save your local delta

```
$ git diff > /tmp/p
```

2) force your local tree to remote master

```
$ git fetch https://github.com/warmcat/libwebsockets +master:m ; git reset --hard m
```

3) reapply your local delta

```
$ patch -p1 --merge < /tmp/p
```

**Comment by lws-team** [comment]
This is getting a bit nuts ^^  We should probably try and see if I can reproduce it on the same OS install.

I installed Ubuntu once before on a VM to find out a different (libuv) problem, I'll look into repeating that.

**Comment by lws-team** [comment]
The mystery at the minute is what's happening here (111897 in your last log)

```
lws_handshake_server
/home/noname/Components/libwebsockets/test-server/test-server-http.c: 802 (lws_http_transaction_completed)
lwsts[14320]: lws_http_transaction_completed: wsi 0x62beb0
```

It's just before the failure which goes on

```
lws_header_table_attach
/home/noname/Components/libwebsockets/lib/parsers.c: 202 (lws_header_table_reset)
_lws_header_table_reset
lws_handshake_server
lwsts[14320]:  forbidding on uri sanitation
```

That last `lws_handshake_server` is supposed to be the start of a new packet with the continuation of the last set of headers.  That means the previous `lws_handshake_server` should be the partial set of headers at the end of the penultimate packet.  But in that case the `wsi->hdr_parsing_completed` should be 0, since we didn't finish parsing those partial headers.  That should disallow the ah being detached.

So it looks like my model of what is happening and where we are in what packet is wrong.  I'm installing your version of Ubutu now it shouldn't take long to try it.

**Comment by lws-team** [comment]
It installed it and built lws , lws works OK with a browser.  But netcat seems different on Ubuntu.

Can I get your packet generator app, or some minimal version of it that creates the problem?

**Comment by rickynelson1939** [comment]
mmm, it written on windows and in Delphi, but it's really simple
it's creating socket, buffer = copy of request enough time so it's length > 2048 * 3
then I create 2 threads in first there is recv loop and in second is loop with send buffer

**Comment by lws-team** [comment]
Ahaha  I reproduced it using a mishmash of BSD and "traditional" netcat.  I can't paste in and out of the VM but here it is for reference.... terminal 1

```
 $ nc -l -k -p 7680 127.0.0.1 | nc 127.0.0.1 7681
```

terminal 2

```
 $ while [ 1 ] ; do echo -n -e "GET /favicon.ico HTTP/1" | nc.traditional -q0 127.0.0.1 7680 ; echo -n -e ".1\x0d\x0a\x0d\x0a" | nc.traditional -q0 127.0.0.1 7680 ; done
```

So...

**Comment by rickynelson1939** [comment]
Wow, really? That's quite a relief actually, I was starting to think it's something wrong with my system or anything else specific to my machine. The level of intrigue is astonishing

**Comment by lws-team** [comment]
Yeah I can reliably reproduce it atm.

It's still a big mystery it can't be reproduced on the thing running the vm, but can be reproduced in the vm.

I'm going to have to do other things soon but this should obviously speed things up when I get back to it.

**Comment by rickynelson1939** [comment]
Okay, no problem. Just don't delete VM if you fix this, cause there is another bug that even I can't reproduce reliably enough (takes ~30 seconds), so...

**Comment by lws-team** [comment]
I moved back up a step towards the root cause, but it's quite puzzling to fit it into something lws is doing.

Basically there are two outcomes possible from running the test I posted earlier.  Here they are with or or two example failure buffers (read with a single `recv()`)

1) The forbidden / header parsing failure we discussed so far.

First example, two contiguous recv()

```
lwsts[18948]: 0000: 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63    .1....GET /favic
lwsts[18948]: 0010: 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D    on.ico HTTP/1.1.
lwsts[18948]: 0020: 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E    ...GET /favicon.
lwsts[18948]: 0030: 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 0D 0A    ico HTTP/1.1....
```

```
lwsts[18948]: 0000: 2E 31 0D 0A 0D 0A                                  .1....  
```

The second recv is missing, or repeating, a packet content.  Weill that could be lws somehow, but then there's the second example, broken inside one atomic recv(), see offset +6

```
lwsts[18948]: 0000: 2E 31 0D 0A 0D 0A 2E 31 0D 0A 0D 0A 47 45 54 20    .1.....1....GET 
lwsts[18948]: 0010: 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54    /favicon.ico HTT
lwsts[18948]: 0020: 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61    P/1.1....GET /fa
lwsts[18948]: 0030: 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31    vicon.ico HTTP/1
lwsts[18948]: 0040: 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63    .1....GET /favic
lwsts[18948]: 0050: 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D    on.ico HTTP/1.1.
lwsts[18948]: 0060: 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E    ...GET /favicon.
lwsts[18948]: 0070: 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 0D 0A    ico HTTP/1.1....
lwsts[18948]: 0080: 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F    GET /favicon.ico
lwsts[18948]: 0090: 20 48 54 54 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54     HTTP/1.1....GET
lwsts[18948]: 00A0: 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54     /favicon.ico HT
lwsts[18948]: 00B0: 54 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66    TP/1.1....GET /f
lwsts[18948]: 00C0: 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F    avicon.ico HTTP/
lwsts[18948]: 00D0: 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69    1.1....GET /favi
lwsts[18948]: 00E0: 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31    con.ico HTTP/1.1
lwsts[18948]: 00F0: 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E    ....GET /favicon
lwsts[18948]: 0100: 2E 69 63 6F 20 48 54 54 50 2F 31                   .ico HTTP/1 
```

There's no ambiguity about the recv buffer start or length AFAICS.  And there is no other connection to the server.

2) Lws hangs up on the client without a message, because he read

`GET /favicon.ico HTTP/1GET /favicon.ico HTTP/1.1`

this is nonsense (again missing, or duplicating a packet) but it's not directly a lexical parser error.

Look at offset 0x3a

```
lwsts[6630]: 0000: 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63    .1....GET /favic
lwsts[6630]: 0010: 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D    on.ico HTTP/1.1.
lwsts[6630]: 0020: 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E    ...GET /favicon.
lwsts[6630]: 0030: 69 63 6F 20 48 54 54 50 2F 31 47 45 54 20 2F 66    ico HTTP/1GET /f
lwsts[6630]: 0040: 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F    avicon.ico HTTP/
lwsts[6630]: 0050: 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69    1.1....GET /favi
lwsts[6630]: 0060: 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 47 45    con.ico HTTP/1GE
lwsts[6630]: 0070: 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48    T /favicon.ico H
lwsts[6630]: 0080: 54 54 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F    TTP/1.1....GET /
lwsts[6630]: 0090: 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50    favicon.ico HTTP
lwsts[6630]: 00A0: 2F 31 2E 31 0D 0A 0D 0A 2E 31 0D 0A 0D 0A 47 45    /1.1.....1....GE
lwsts[6630]: 00B0: 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48    T /favicon.ico H
lwsts[6630]: 00C0: 54 54 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F    TTP/1.1....GET /
lwsts[6630]: 00D0: 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50    favicon.ico HTTP
lwsts[6630]: 00E0: 2F 31 2E 31 0D 0A 0D 0A                            /1.1....  
```

Maybe I am missing the point but actually I don't see how lws can cause these missing or repeated packets.  Consider the examples where it's broken inside a single recv()... there is only one thread in lws.  If something else was reading from the socket, it would have read the pending data captured by the recv first.... the error is at +0x3a in the read in one case.  lws is singlethreaded too... it can't be competing with itself during the atomic `recv()` syscall.

Maybe it's something to do with the netcat packet generation, but I can run more or less the same thing on Fedora (its netcat seems a bit more advanced than the ones on Ubuntu) for as long as I like (I tried up to 15m) with not once any symptom.  And it's suspicious you can create the symptom using a completely unrelated method (and platform).

Since lws test apps are static-linked by default I copied the Ubuntu test app out of the vm and into the parent Fedora instance.  He has no problems (I ran him for 10m) running on Fedora, but he is spewing a lot of logs and he doesn't seem to coalesce much.  This machine is pretty fast I guess lws can keep up with the netcat process loop in bash.

The Ubuntu-built test server is mostly static linked but does have some imports, linux-vdso, libc.so.6 and ld-linux.  So it could be a problem with glibc (ie, `recv()` provider) on the Ubuntu instance that doesn't exist on the Fedora one.

Any ideas so I can avoid embarrassing myself with a wrong best guess outside of lws welcome.

**Comment by rickynelson1939** [comment]
recv is returning buffers content of which make no sense? I can't remember anything like it in my recv_dump's so it's quite new, it's hard for me to bash scripts, but... ok, maybe I just write a program for Ubuntu that is reproducing the bug so we can make sure there is everything ok on client side?

**Comment by lws-team** [comment]
Sure... what the bash stuff does is first listen on 7680 and establish one client connection on 7681, to the test server.  Then it sits in a loop firing a GET request fragmented into two parts at 7680, which passes through anything it receives to its connection to 7681.

Although that works fine on Fedora, on Ubuntu sometimes what we get on 7681 has one fragment either missing or repeated, generating the two symptoms, depending on which fragment.  It's possible that behaviour is coming from the netcat setup somehow, rather than lws.  So having a different fragment generator we can both use the same won't hurt.

**Comment by lws-team** [comment]
In the meanwhile... 

1) I found switches to control nc in and out buffer size, and set these to 160KB, it makes no difference.  This is aimed at the idea the listening netcat drops the fragment because we spam it.

2) I split the test header into three so we can tell if it is dropping or repeating (it's the same result either way with only two pieces), something is **dropping** one fragment (which fragment seems random) to make the problem.  Since `recv()` just gives us the coalesced buffer atomically, and lws nor `recv()` have any idea where the original fragments begin and end, this must be happening outside of lws (as you say, can be on packet generation side).

3) I also amended the nc loop to test for errors from each nc and bail, the first error is after lws hangs up on us, the listen nc exited and it reports "Connection Refused".  This is aimed at the idea something goes wrong with delivering the fragment to the listening netcat but we miss it in the resulting spew of Connection Refused.

4) Adding a 50ms sleep after each fragment send removes the packet coalescing.... lws receives a buffer consisting of one fragment each time.  The problem didn't come in at least 10 minutes then.

5) I copied the Fedora-built test server binary to the Ubuntu vm - he shows the same failing behaviours as the Ubuntu-built one.  So it is not to do with the toolchain as far as generation of the lws binary goes.

**Comment by rickynelson1939** [comment]
I code my test program in C, executing it on the same machine gives no bug :(
Maybe you'll have some luck with it, it's basically the same program ported to Linux, so it's strange that is does not trigger the bug
[LWS_PIPELINE_TEST.zip](https://github.com/warmcat/libwebsockets/files/1110495/LWS_PIPELINE_TEST.zip)

**Comment by rickynelson1939** [comment]
Oh my god, if my test is running on Ubuntu, even my Windows test can't trigger the bug, it only triggers after I cancel Ubuntu test program

**Comment by lws-team** [comment]
I didn't try your app yet, but the bug definitely requires packets to end partway through the http request to notice it.  Otherwise dropping a packet would just drop one or more whole requests, and you would have no observable symptom from that.

**Comment by rickynelson1939** [comment]
```
int dump_num = 0;

#if LWS_POSIX
LWS_VISIBLE int
lws_ssl_capable_read_no_ssl(struct lws *wsi, unsigned char *buf, int len)
{
	struct lws_context *context = wsi->context;
	struct lws_context_per_thread *pt = &context->pt[(int)wsi->tsi];
	int n;
	char dump_filename[256];
	FILE *dump_file;

	lws_stats_atomic_bump(context, pt, LWSSTATS_C_API_READ, 1);

	n = recv(wsi->desc.sockfd, (char *)buf, len, 0);

	// assert(n == 2048);

	sprintf(dump_filename, "/home/noname/Components/libwebsockets/build/recv_dump/%d.txt", dump_num);

	dump_file = fopen(dump_filename, "w");
  	assert(fwrite(buf, sizeof(char), n, dump_file) == n);

	/*
	dump_file = fopen(dump_filename, "r");
	n = fread(buf, sizeof(char), 2048, dump_file);
	*/

  	fclose(dump_file);

  	dump_num++;
```
[recv_dump_new.zip](https://github.com/warmcat/libwebsockets/files/1110519/recv_dump_new.zip)

Here's the dump right before bug and dump code, but there seems a completely legit input, or maybe I overlooked something?

**Comment by lws-team** [comment]
This blew a "forbidden"?

**Comment by rickynelson1939** [comment]
Yes

**Comment by lws-team** [comment]
Welp I caught my netcat stuff sending the already-broken packets to lws using tcpdump.

```
08:50:36.421436 IP localhost.44796 > localhost.7681: Flags [P.], seq 19627:19765, ack 1011297, win 32742, options [nop,nop,TS val 15611226 ecr 15611226], length 138
	0x0000:  4500 00be a4f1 4000 4006 9746 7f00 0001  E.....@.@..F....
	0x0010:  7f00 0001 aefc 1e01 9c27 dace dbcc bede  .........'......
	0x0020:  8018 7fe6 feb2 0000 0101 080a 00ee 355a  ..............5Z
	0x0030:  00ee 355a 6f20 4854 5450 2f31 2e31 0d0a  ..5Zo.HTTP/1.1..
	0x0040:  0d0a 2e31 0d0a 0d0a 4745 5420 2f66 6176  ...1....GET./fav
	0x0050:  6963 6f6e 2e69 636f 2048 5454 502f 316f  icon.ico.HTTP/1o
	0x0060:  2048 5454 502f 312e 310d 0a0d 0a47 4554  .HTTP/1.1....GET
	0x0070:  202f 6661 7669 636f 6e2e 6963 6f20 4854  ./favicon.ico.HT
	0x0080:  5450 2f31 2e31 0d0a 0d0a 2e31 0d0a 0d0a  TP/1.1.....1....
	0x0090:  4745 5420 2f66 6176 6963 6f6e 2e69 636f  GET./favicon.ico
	0x00a0:  2048 5454 502f 3147 4554 202f 6661 7669  .HTTP/1GET./favi
	0x00b0:  636f 6e2e 6963 6f20 4854 5450 2f31       con.ico.HTTP/1
```

```
lwsts[29321]: 0000: 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 0D 0A 2E 31    o HTTP/1.1.....1
lwsts[29321]: 0010: 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E    ....GET /favicon
lwsts[29321]: 0020: 2E 69 63 6F 20 48 54 54 50 2F 31 6F 20 48 54 54    .ico HTTP/1o HTT
lwsts[29321]: 0030: 50 2F 31 2E 31 0D 0A 0D 0A 47 45 54 20 2F 66 61    P/1.1....GET /fa
lwsts[29321]: 0040: 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31    vicon.ico HTTP/1
lwsts[29321]: 0050: 2E 31 0D 0A 0D 0A 2E 31 0D 0A 0D 0A 47 45 54 20    .1.....1....GET 
lwsts[29321]: 0060: 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54    /favicon.ico HTT
lwsts[29321]: 0070: 50 2F 31 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E    P/1GET /favicon.
lwsts[29321]: 0080: 69 63 6F 20 48 54 54 50 2F 31                      ico HTTP/1   
```

So it looks like the "reproduction" is unrelated to yours... which puts us back to square 1.

**Comment by lws-team** [comment]
So the end result of your C test app was that it can't reproduce it either?

What exactly is the setup that does reproduce it.... Ubuntu is in a vm on Windows?

**Comment by rickynelson1939** [comment]
Windows 7 <-> Ubuntu 16.04, two physical machines
connected via LAN cable

**Comment by rickynelson1939** [comment]
maybe it's worth trying to send packets from Fedora -> Ubuntu, or even from my machine to your virtual Ubuntu, don't know if this possible though

**Comment by lws-team** [comment]
You should be able to touch my Fedora's test server at home.warmcat.com:7682

**Comment by lws-team** [comment]
Oh it blew!

**Comment by rickynelson1939** [comment]
did the forbidden happen?

**Comment by lws-team** [comment]
Yes... on Fedora...

```
lwsts[10249]: lws_http_transaction_completed: wsi 0x1ceeea0
lwsts[10249]: lws_http_transaction_completed: wsi 0x1ceeea0
lwsts[10249]: lws_http_transaction_completed: wsi 0x1ceeea0
lwsts[10249]:  forbidding on uri sanitation
```

Can you set it up so it's constantly trying to connect and run in a loop?

**Comment by rickynelson1939** [comment]
okay, I'll do it now

**Comment by lws-team** [comment]
I brought over the debug logging changes from Ubuntu, have to see if that affects it as you found logging could make it go away.

**Comment by rickynelson1939** [comment]
Should sending stuff to your server and reconnect now

**Comment by lws-team** [comment]
Yes it seems to work thanks... I removed some hexdumps and it could make the problem again after a few seconds

```
GET /favicon.ico HTTP/1.1
Connection: keep-alive
Host: 192.168.137.27

GET /favicon.ico HTTP/1.1
Connection: keep-al
lwsts[11345]: *buf on fail 0x18dacd2
lwsts[11345]: 0000: 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20    .1..Connection: 
lwsts[11345]: 0010: 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74    keep-alive..Host
lwsts[11345]: 0020: 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37    : 192.168.137.27
lwsts[11345]: 0030: 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E    ....GET /favicon
lwsts[11345]: 0040: 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43    .ico HTTP/1.1..C
lwsts[11345]: 0050: 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D    onnection: keep-
lwsts[11345]: 0060: 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32    alive..Host: 192
lwsts[11345]: 0070: 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47    .168.137.27....G
lwsts[11345]: 0080: 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20    ET /favicon.ico 
lwsts[11345]: 0090: 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63    HTTP/1.1..Connec
lwsts[11345]: 00A0: 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65    tion: keep-alive
lwsts[11345]: 00B0: 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E    ..Host: 192.168.
lwsts[11345]: 00C0: 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66    137.27....GET /f
lwsts[11345]: 00D0: 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F    avicon.ico HTTP/
lwsts[11345]: 00E0: 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A    1.1..Connection:
lwsts[11345]: 00F0: 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73     keep-alive..Hos
lwsts[11345]: 0100: 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32    t: 192.168.137.2
lwsts[11345]: 0110: 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F    7....GET /favico
lwsts[11345]: 0120: 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A    n.ico HTTP/1.1..
lwsts[11345]: 0130: 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70    Connection: keep
lwsts[11345]: 0140: 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39    -alive..Host: 19
lwsts[11345]: 0150: 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A    2.168.137.27....
lwsts[11345]: 0160: 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F    GET /favicon.ico
lwsts[11345]: 0170: 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65     HTTP/1.1..Conne
lwsts[11345]: 0180: 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76    ction: keep-aliv
lwsts[11345]: 0190: 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36 38    e..Host: 192.168
lwsts[11345]: 01A0: 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F    .137.27....GET /
lwsts[11345]: 01B0: 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50    favicon.ico HTTP
lwsts[11345]: 01C0: 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E    /1.1..Connection
lwsts[11345]: 01D0: 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F    : keep-alive..Ho
lwsts[11345]: 01E0: 73 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E    st: 192.168.137.
lwsts[11345]: 01F0: 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63    27....GET /favic
lwsts[11345]: 0200: 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D    on.ico HTTP/1.1.
lwsts[11345]: 0210: 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65    .Connection: kee
lwsts[11345]: 0220: 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31    p-alive..Host: 1
lwsts[11345]: 0230: 39 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D    92.168.137.27...
lwsts[11345]: 0240: 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63    .GET /favicon.ic
lwsts[11345]: 0250: 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E    o HTTP/1.1..Conn
lwsts[11345]: 0260: 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69    ection: keep-ali
lwsts[11345]: 0270: 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36    ve..Host: 192.16
lwsts[11345]: 0280: 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20    8.137.27....GET 
lwsts[11345]: 0290: 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54    /favicon.ico HTT
lwsts[11345]: 02A0: 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F    P/1.1..Connectio
lwsts[11345]: 02B0: 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48    n: keep-alive..H
lwsts[11345]: 02C0: 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37    ost: 192.168.137
lwsts[11345]: 02D0: 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69    .27....GET /favi
lwsts[11345]: 02E0: 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31    con.ico HTTP/1.1
lwsts[11345]: 02F0: 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65    ..Connection: ke
lwsts[11345]: 0300: 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20    ep-alive..Host: 
lwsts[11345]: 0310: 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A    192.168.137.27..
lwsts[11345]: 0320: 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69    ..GET /favicon.i
lwsts[11345]: 0330: 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E    co HTTP/1.1..Con
lwsts[11345]: 0340: 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C    nection: keep-al
lwsts[11345]: 0350: 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31    ive..Host: 192.1
lwsts[11345]: 0360: 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54    68.137.27....GET
lwsts[11345]: 0370: 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54     /favicon.ico HT
lwsts[11345]: 0380: 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69    TP/1.1..Connecti
lwsts[11345]: 0390: 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A    on: keep-alive..
lwsts[11345]: 03A0: 48 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E 31 33    Host: 192.168.13
lwsts[11345]: 03B0: 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76    7.27....GET /fav
lwsts[11345]: 03C0: 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E    icon.ico HTTP/1.
lwsts[11345]: 03D0: 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B    1..Connection: k
lwsts[11345]: 03E0: 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A    eep-alive..Host:
lwsts[11345]: 03F0: 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D     192.168.137.27.
lwsts[11345]: 0400: 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E    ...GET /favicon.
lwsts[11345]: 0410: 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F    ico HTTP/1.1..Co
lwsts[11345]: 0420: 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61    nnection: keep-a
lwsts[11345]: 0430: 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E    live..Host: 192.
lwsts[11345]: 0440: 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45    168.137.27....GE
lwsts[11345]: 0450: 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48    T /favicon.ico H
lwsts[11345]: 0460: 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74    TTP/1.1..Connect
lwsts[11345]: 0470: 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D    ion: keep-alive.
lwsts[11345]: 0480: 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E 31    .Host: 192.168.1
lwsts[11345]: 0490: 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61    37.27....GET /fa
lwsts[11345]: 04A0: 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31    vicon.ico HTTP/1
lwsts[11345]: 04B0: 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20    .1..Connection: 
lwsts[11345]: 04C0: 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74    keep-alive..Host
lwsts[11345]: 04D0: 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37    : 192.168.137.27
lwsts[11345]: 04E0: 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E    ....GET /favicon
lwsts[11345]: 04F0: 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43    .ico HTTP/1.1..C
lwsts[11345]: 0500: 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D    onnection: keep-
lwsts[11345]: 0510: 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32    alive..Host: 192
lwsts[11345]: 0520: 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47    .168.137.27....G
lwsts[11345]: 0530: 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20    ET /favicon.ico 
lwsts[11345]: 0540: 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63    HTTP/1.1..Connec
lwsts[11345]: 0550: 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65    tion: keep-alive
lwsts[11345]: 0560: 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E    ..Host: 192.168.
lwsts[11345]: 0570: 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66    137.27....GET /f
lwsts[11345]: 0580: 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F    avicon.ico HTTP/
lwsts[11345]: 0590: 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A    1.1..Connection:
lwsts[11345]: 05A0: 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73     keep-alive..Hos
lwsts[11345]: 05B0: 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32    t: 192.168.137.2
lwsts[11345]: 05C0: 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F    7....GET /favico
lwsts[11345]: 05D0: 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A    n.ico HTTP/1.1..
lwsts[11345]: 05E0: 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70    Connection: keep
lwsts[11345]: 05F0: 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31 39    -alive..Host: 19
lwsts[11345]: 0600: 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A    2.168.137.27....
lwsts[11345]: 0610: 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63 6F    GET /favicon.ico
lwsts[11345]: 0620: 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65     HTTP/1.1..Conne
lwsts[11345]: 0630: 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69 76    ction: keep-aliv
lwsts[11345]: 0640: 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36 38    e..Host: 192.168
lwsts[11345]: 0650: 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F    .137.27....GET /
lwsts[11345]: 0660: 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54 50    favicon.ico HTTP
lwsts[11345]: 0670: 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E    /1.1..Connection
lwsts[11345]: 0680: 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F    : keep-alive..Ho
lwsts[11345]: 0690: 73 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E    st: 192.168.137.
lwsts[11345]: 06A0: 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69 63    27....GET /favic
lwsts[11345]: 06B0: 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31 0D    on.ico HTTP/1.1.
lwsts[11345]: 06C0: 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65 65    .Connection: kee
lwsts[11345]: 06D0: 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20 31    p-alive..Host: 1
lwsts[11345]: 06E0: 39 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A 0D    92.168.137.27...
lwsts[11345]: 06F0: 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69 63    .GET /favicon.ic
lwsts[11345]: 0700: 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E 6E    o HTTP/1.1..Conn
lwsts[11345]: 0710: 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C 69    ection: keep-ali
lwsts[11345]: 0720: 76 65 0D 0A 48 6F 73 74 3A 20 31 39 32 2E 31 36    ve..Host: 192.16
lwsts[11345]: 0730: 38 2E 31 33 37 2E 32 37 0D 0A 0D 0A 47 45 54 20    8.137.27....GET 
lwsts[11345]: 0740: 2F 66 61 76 69 63 6F 6E 2E 69 63 6F 20 48 54 54    /favicon.ico HTT
lwsts[11345]: 0750: 50 2F 31 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F    P/1.1..Connectio
lwsts[11345]: 0760: 6E 3A 20 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48    n: keep-alive..H
lwsts[11345]: 0770: 6F 73 74 3A 20 31 39 32 2E 31 36 38 2E 31 33 37    ost: 192.168.137
lwsts[11345]: 0780: 2E 32 37 0D 0A 0D 0A 47 45 54 20 2F 66 61 76 69    .27....GET /favi
lwsts[11345]: 0790: 63 6F 6E 2E 69 63 6F 20 48 54 54 50 2F 31 2E 31    con.ico HTTP/1.1
lwsts[11345]: 07A0: 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20 6B 65    ..Connection: ke
lwsts[11345]: 07B0: 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74 3A 20    ep-alive..Host: 
lwsts[11345]: 07C0: 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37 0D 0A    192.168.137.27..
lwsts[11345]: 07D0: 0D 0A 47 45 54 20 2F 66 61 76 69 63 6F 6E 2E 69    ..GET /favicon.i
lwsts[11345]: 07E0: 63 6F 20 48 54 54 50 2F 31 2E 31 0D 0A 43 6F 6E    co HTTP/1.1..Con
lwsts[11345]: 07F0: 6E 65 63 74 69 6F 6E 3A 20 6B 65 65 70 2D 61 6C    nection: keep-al
```

It's the same place as in packet 82 in your last dump.

**Comment by lws-team** [comment]
Hmmmm so just before it blows,

```
lwsts[11812]: parsing complete: 'GET /favicon.ico HTTP/1.1
Connection: keep-alive
Host: 192.168.137.27

', rxpos 1973 rxlen 2048
lwsts[11812]: lws_handshake_server: lws_parse sees parsing complete
lwsts[11812]: lws_read: exited lws_handshake_server
lwsts[11812]: lws_server_socket_service: wsi 0xe38ea0: ah read rxpos 2048, rxlen 2048, wsi->hdr_parsing_completed 1
```

The first line about parsing complete is the parsed characters copied into a side buffer and dumped when the parsing was finished.  rxpos at the start of that was +1973 in the 2048-byte rx buffer.

The parsed chars total 75 chars, so it correctly advances rxpos to +2048, ie, it has exhausted the buffer.  Everything about that is OK.

But that means the penultimate packet we got from you ended exactly on the end of a header set.

So the first chars in the next packet should be "GET".  But they aren't, the last packet starts

lwsts[11812]: 0000: 2E 31 0D 0A 43 6F 6E 6E 65 63 74 69 6F 6E 3A 20    .1..Connection: 
lwsts[11812]: 0010: 6B 65 65 70 2D 61 6C 69 76 65 0D 0A 48 6F 73 74    keep-alive..Host
lwsts[11812]: 0020: 3A 20 31 39 32 2E 31 36 38 2E 31 33 37 2E 32 37    : 192.168.137.27

Either we dropped or swallowed a packet in lws or your Delphi app has missed out sending a packet...

**Comment by rickynelson1939** [comment]
It's strange, since my app buffer is consist of duped complete requests and it just goes to send (6300 bytes, 75 bytes per request, so 84 requests per buffer (per `send` call)), of course it may be Windows who dropped something, but... let's just save it for later, how about that my recv_dumps contain a valid traffic?

**Comment by lws-team** [comment]
I think the problem is looking like the "Either we dropped or swallowed a packet in lws" part atm

**Comment by rickynelson1939** [comment]
Maybe it's good idea to dump what's come from recv, just in case there is actually a traffic problem

**Comment by lws-team** [comment]
I've copied the previous buffer into a side buffer just before reading the next... it's the correct buffer in there.

We seem to read the buffer twice at the point it blows, losing it.

**Comment by lws-team** [comment]
I pushed a patch on master which I think nails it

https://github.com/warmcat/libwebsockets/commit/ff9a24de1cd7017ead6d0e50bec1fda09157cc9b

**Comment by rickynelson1939** [comment]
Oh my, I think you fix it. I'm just gonna write validation of whatever comes from recv on test app side to be sure.

**Comment by rickynelson1939** [comment]
Everything seems to be working, congratulations :)
Thank you very much for fixing such a complex bug. Is debugging revealed why it was so hard to reproduce it?

**Comment by lws-team** [comment]
It depended on whether the next packet had arrived or not, and the packet just finished had to end on a headers boundary.  If the next packet hadn't come until later, there's no problem the ah is dropped.  When the packet came later it was reacquired and the new packet started on a new header set and everything OK.

If the next packet had arrived, we read it immediately after .  But we ran intro trouble then because we had been at the end of the rx buffer, and we had completed parsing.  It felt it was OK to drop the ah, this time with the consequence of losing the rx buffer.  It reacquired the ah at the next buffer arrival, and read the next buffer -- but now it was out of sync having dropped the buffer it read earlier.  The patch defeats that misunderstanding.

Anyway thanks for the help reproducing it.  I take it the other bug you mentioned went away as well?

**Comment by rickynelson1939** [comment]
Well, I have to try it now, but I have a feeling it haven't, although who knows...
the thing was like this: if I spam http requests that cause response with `lws_return_http_status` then sometimes (might be 30 minutes without a bug, but might be in a first second) there was an assert telling me that I did write outside of writable callback

**Comment by rickynelson1939** [comment]
yeap, catch it right now...

**Comment by rickynelson1939** [comment]
__GI_raise() at raise.c:54 0x7ffff7a42428	
__GI_abort() at abort.c:89 0x7ffff7a4402a	
__assert_fail_base() at assert.c:92 0x7ffff7a3abd7	
__GI___assert_fail() at assert.c:101 0x7ffff7a3ac82	
lws_issue_raw() at output.c:126 0x40d193	
lws_write() at output.c:533 0x40dcdf	
lws_return_http_status() at header.c:256 0x414044	
callback_http() at test-server-http.c:273 0x4056dc	
lws_http_action() at server.c:956 0x417522	
lws_handshake_server() at server.c:1,461 0x41873e

**Comment by lws-team** [comment]
OK... I think it makes sense to open a new issue for that then.  Or we might blow up github on this one.
